File: 1-Two-Sum-Problem/twoSum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        valueMap = {}    # value : index
        for index, value in enumerate(nums):
            diff = target - value
            if diff in valueMap:        # if difference found
                return [valueMap[diff], index]
            valueMap[value] = index     # if difference not found
        return None     # no solution

File: 1-Two-Sum-Problem/twoSum.java
/*
 * Algorithm:
 * We are iterating and inserting elements into the hash table, 
 * we also look back to check if current element's complement already 
 * exists in the hash table. If it exists, we have found a solution 
 * and return the indices immediately.
 * Time complexity: O(n)
 * Space complexity: O(n)
*/

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> valueMap = new HashMap<>(); // value : index
        for (int i = 0; i < nums.length; ++i) {
            int diff = target - nums[i];
            if (valueMap.containsKey(diff)) {   // if difference found
                return new int[] { valueMap.get(diff), i};
            }
            valueMap.put(nums[i], i);   // if difference not found
        }
        return null; // no solution
    }
}